### PR TITLE
[stubsabot] add script

### PR DESCRIPTION
**Pull request by hauntsaninja**
This implements the suggestion in https://github.com/python/typeshed/pull/7537#issuecomment-1075617617 (and builds on some of the code from that PR). The proposed workflow here is we automatically get PRs updating the version in METADATA.toml. If stubtest passes, we can move on with our lives with reasonable confidence the stubs are okay (especially once we get rid of `--ignore-missing-stub` for third party stubs).

I haven't yet tested it end to end or tried getting it to work inside Github Actions, figured this was about the right point in time to discuss.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Great idea!

File: scripts/stubsabot.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+def _check_spec(updated_spec: str, version: packaging.version.Version) -> str:
+    assert version in packaging.specifiers.SpecifierSet("==" + updated_spec), f"{version} not in {updated_spec}"
+    return updated_spec
+
+
+def get_updated_version_spec(spec: str, version: packaging.version.Version) -> str:
+    if not spec.endswith(".*"):
+        return _check_spec(version.base_version, version)
+
+    specificity = spec.count(".") if spec.removesuffix(".*") else 0
```
[the remark this answers is not in the document]
Comment (hauntsaninja) [reply scripts/stubsabot.py:81]: Good point, auto-updating to a py.typed version is suboptimal.

Not sure that major version is the right heuristic (I've seen several packages introduce py.typed in minor versions and most major version changes won't add py.typed). Maybe we should just download the package, check for py.typed and obsolete the package if so.

**Comment by hauntsaninja** [comment]
Okay, added automatic obsoletion to this. I also did a one time pass to manually make PRs for stubs that are obsolete, so that we get the correct version numbers.

**Comment by hauntsaninja** [comment]
Thoughts on next steps here? My plan is to open small batches of these so people get a sense for the workflow, but let me know if stubsabot is annoying and I won't.

**Comment by JelleZijlstra** [comment]
Sounds like a good plan.

**Comment by AlexWaygood** [comment]
What should we do for stubsabot PRs like #8081, where it's clear that a little bit of work needs to be done to update the stub? Should we leave those PRs open until one of us has time to do the work? Close the PR, and open an issue to remind us that the stubs for that package are out of date?

**Comment by hauntsaninja** [comment]
One possible workflow:

- Leave the PR open as a reminder. Doing so will also prevent the current stubsabot code from opening another PR. I don't have much experience with dependsabot, but I think folks commonly leave those open for a bit as well.
- When one of us has the time to fix it, we just open another PR, merge, and close the stubsabot PR.

Feels like that could be pretty manageable in the steady state (not unlike the daily stubtest when people release new versions that match the existing specifier in METADATA.toml, but break stubtest). And the current backlog we can handle on our own time, either by not running stubsabot in CI for a while or by using an allowlist or something.

**Comment by AlexWaygood** [comment]
> * Leave the PR open as a reminder. Doing so will also prevent the current stubsabot code from opening another PR. I don't have much experience with dependsabot, but I think folks commonly leave those open for a bit as well.

If you close a Dependabot PR, he (she? it?) folornly leaves a message saying he won't contact you again until there's a new release, e.g. https://github.com/python/devguide/pull/784#issuecomment-1023619519. (Not saying I want that level of advanced automation from stubsabot, just for reference.)

> or by using an allowlist or something.

I think I prefer the idea of an allowlist. I'm not a _massive_ fan of having a large number of open PRs with red crosses next to them, but an allowlist we can work through in our own time. It would also be a nice thing to point contributors towards in `CONTRIBUTING.md`. (Want to help out? Here's a handy list of all our stubs which we know to be somewhat out of date!)

Outside contributors might have a hard time contributing towards stubsabot PRs, on the other hand, since they won't be able to push towards stubsabot's branch. (They can open a new PR, of course, but that might not be obvious.)

**Comment by hauntsaninja** [comment]
Okay great. Updated plan:
- The basic script is ready for review and merge. Merging now will make review of future tweaks easier.
- I'll continue to manually open small batches of these so we can see what other workflow issues crop up. I'll make sure that we have at most three open stubsabot PRs failing CI.
- I'm hopeful we can close the backlog quickly, if that proves not to be the case and we want to make this run automatically, we'll use an allowlist for the backlog.
- At some point, make this run in CI (probably on a weekly basis).

**Comment by AlexWaygood** [comment]
> Okay great. Updated plan:
> 
> * The basic script is ready for review and merge. Merging now will make review of future tweaks easier.
> * I'll continue to manually open small batches of these so we can see what other workflow issues crop up. I'll make sure that we have at most three open stubsabot PRs failing CI.
> * I'm hopeful we can close the backlog quickly, if that proves not to be the case and we want to make this run automatically, we'll use an allowlist for the backlog.
> * At some point, make this run in CI (probably on a weekly basis).

Okay great, I'm on board — thanks for doing this!

File: scripts/stubsabot.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+import packaging.specifiers
+import packaging.version
+import tomli
+import tomlkit
+
+
+class ActionLevel(enum.IntEnum):
+    nothing = 0  # make no changes
+    local = 1  # make changes that affect local repo
+    everything = 2  # do everything, e.g. open PRs
```
Comment (AlexWaygood) [inline scripts/stubsabot.py:31]: Very minor, but my favourite pattern for things like this is to do something like `http.HTTPStatus`. It allows you to have per-member docstrings:

```suggestion
class ActionLevel(IntEnum):
    def __new__(cls, value: int, doc: str):
        member = int.__new__(cls, value)
        member._value_ = value
        member.__doc__ = doc
        return member

    nothing = 0, "make no changes"
    local = 1, "make changes that affect local repo"
    everything = 2, "do everything, e.g. open PRs"
```

**Review by AlexWaygood — COMMENTED** [review]
A few nits, but this basically looks great! (I haven't tried the script out myself, but it all looks reasonably self-explanatory, and the PRs you've been posting look solid.)

**Comment by hauntsaninja** [comment]
Thanks! Merged to make review of future changes easier, because I named my branch poorly and it conflicts with the prefix in my fork and since this seems like a passable prototype. Still very interested in feedback (and not attached to doing this if people have concerns), so feel free to post here or open an issue for follow ups as they come up :-)

**Pull request merged**